src/Hopper/Internal/LoweredCore/EvalClosureConvertedANF.hs

```diff
@@ -260,6 +283,42 @@ applyCC symbolReg envStk stack alloc = case alloc of
   FunAppCC var vec -> enterClosureCC symbolReg envStk stack (var, vec)
   PrimAppCC opId vec -> enterPrimAppCC symbolReg envStk stack (opId, vec)
 
+{- | enterClosureCC has to resolve its first heap ref argument to the closure code id
+and then it pushes
+-}
+-- TODO: reduce duplication between lookupHeap{Closure,Thunk}.
+--       e.g. logic for whether env is "compatible"
+enterClosureCC
+  :: SymbolRegistryCC
+  -> EnvStackCC
+  -> ControlStackCC
+  -> (Variable, V.Vector Variable)
+  -> forall c. EvalCC c s (V.Vector Ref)
+enterClosureCC _symbolReg _env stack (GlobalVarSym gvsym, _) = do
+  errMsg <- return $
+    "`enterClosureCC` expected a local ref, received a global: "++ show gvsym
+  throwEvalError (\step -> PanicMessageConstructor(stack, 0, InterpreterStepCC step, errMsg))
+enterClosureCC symbolReg env stack (LocalVar localvar, args) = do
```

Comment:
I wouldn't mind doing the rest of the changes we've discussed myself this
evening and landing it.  Need some concrete coding to get the fly wheels
spinning

On Thursday, April 7, 2016, Joel Burget <notifications@github.com> wrote:

> In src/Hopper/Internal/LoweredCore/EvalClosureConvertedANF.hs
> <https://github.com/hopper-lang/hopper/pull/4#discussion_r58878284>:
>
> > +{- | enterClosureCC has to resolve its first heap ref argument to the closure code id
> > +and then it pushes
> > +-}
> > +-- TODO: reduce duplication between lookupHeap{Closure,Thunk}.
> > +--       e.g. logic for whether env is "compatible"
> > +enterClosureCC
> > +  :: SymbolRegistryCC
> > +  -> EnvStackCC
> > +  -> ControlStackCC
> > +  -> (Variable, V.Vector Variable)
> > +  -> forall c. EvalCC c s (V.Vector Ref)
> > +enterClosureCC _symbolReg _env stack (GlobalVarSym gvsym, _) = do
> > +  errMsg <- return $
> > +    "`enterClosureCC` expected a local ref, received a global: "++ show gvsym
> > +  throwEvalError (\step -> PanicMessageConstructor(stack, 0, InterpreterStepCC step, errMsg))
> > +enterClosureCC symbolReg env stack (LocalVar localvar, args) = do
>
> This is wrong. We should use args. Still need to fix this up.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly or view it on GitHub
> <https://github.com/hopper-lang/hopper/pull/4/files/0f5a3956ca74372e8505940b93959d62d7b8e74f#r58878284>
>